scripts/partition_manager.py

```diff
@@ -244,23 +241,139 @@ def set_shared_size(reqs, sub_partitions, total_size):
 
+        if align_if_required(i, dynamic_partitions, True, reqs, solution):
+            _set_addresses_and_align(reqs, sub_partitions, solution, size, start, dynamic_partitions)
+
+    for i in range(len(solution) - 1, solution.index('app'), -1):
+        current = solution[i]
+
+        if i == len(solution) - 1:
+            reqs[current]['address'] = size - reqs[current]['size']
+        else:
+            higher_partition = solution[i + 1]
+            reqs[current]['address'] = reqs[higher_partition]['address'] - reqs[current]['size']
+
+        if align_if_required(i, dynamic_partitions, False, reqs, solution):
+            _set_addresses_and_align(reqs, sub_partitions, solution, size, start, dynamic_partitions)
+
+
+def align_if_required(i, dynamic_partitions, move_up, reqs, solution):
+    current = solution[i]
+    if 'placement' in reqs[current] and 'align' in reqs[current]['placement']:
+        required_offset = align_partition(current, reqs, move_up, dynamic_partitions)
+        if required_offset:
+            solution_index = i if move_up else i + 1
+            solution.insert(solution_index, required_offset)
+            return True
+    return False
+
+
+def align_partition(current, reqs, move_up, dynamic_partitions):
+    required_offset = get_required_offset(align=reqs[current]['placement']['align'], start=reqs[current]['address'],
+                                          size=reqs[current]['size'], move_up=move_up)
+    if not required_offset:
+        return None
+
+    empty_partition_size = required_offset
+
+    if current not in dynamic_partitions:
+        if move_up:
+            empty_partition_address = reqs[current]['address']
+            reqs[current]['address'] += required_offset
+        else:
+            empty_partition_address = reqs[current]['address'] + reqs[current]['size']
+            reqs[current]['address'] -= required_offset
+    elif not move_up:
+        empty_partition_address, empty_partition_size = \
+            align_dynamic_partition(dynamic_partitions, current, reqs, required_offset)
+    else:
+        raise RuntimeError("Invalid combination, can not have dynamic partition in front of app with alignment")
+
+    e = 'EMPTY_{}'.format(len([x for x in reqs.keys() if 'EMPTY' in x]))
+    reqs[e] = {'address': empty_partition_address,
+               'size': empty_partition_size,
+               'placement': {'before' if move_up else 'after': [current]}}
+
+    if current not in dynamic_partitions:
+        # We have stolen space from the 'app' partition. Hence, all partitions which share size with 'app' partition
+        # must have their sizes reduced. Note that the total amount of 'stealing' is divided between the partitions
+        # sharing size with app (including 'app' itself).
+        for p in dynamic_partitions:
+            reqs[p]['size'] = reqs[p]['size'] - (reqs[e]['size'] // len(dynamic_partitions))
```

Comment:
AFAICS this is not a problem since we solve from start -> app and then from end -> app.

Only the start addresses **after** the dynamic area (we only support one logical dynamic area) will get new start addresses. When placing the partitions after the dynamic area, we iterate from the end, injecting empty partitions behind the partition being aligned. That way there is no need to adjust the start addresses of the later partitions at any point.